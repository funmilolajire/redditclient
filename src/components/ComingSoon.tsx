import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';

export const ComingSoon = () => {
    const boxRef = useRef<HTMLDivElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({ repeatDelay: 2 })
        tl.from([boxRef.current], {
            y: '-55vh',
            duration: 3,
            delay: 2,
            opacity: 0,
            ease: 'elastic'
        })
    })

    return (
        <div ref={boxRef} className="container flex items-center justify-center h-screen font-extrabold ComingSoon">
            <div className="bg-yellow-500 border-2 border-gray-800 border-solid skew-y-2 transform-gpu text-gray-50 text-9xl">
                <p>Coming Soon</p>
            </div>
        </div>
    )
}