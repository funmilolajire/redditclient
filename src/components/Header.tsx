import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';
import logo from '../images/logo192.png';

export const Header = () => {
    const boxRef = useRef<HTMLDivElement>(null);
    const logoRef = useRef<HTMLImageElement>(null);
    const titleRef = useRef<HTMLHeadingElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({});
        tl.from([boxRef.current], {
            duration: 1
        })
        tl.from([logoRef.current], {
            x: '-3000px',
            opacity: 0,
            ease: 'circ'
        })
        tl.from([titleRef.current], {
            x: '3000px',
            opacity: 0,
            ease: 'slow'
        })
    })

    return (
        <header>
            <div className="header-inner">
                <section ref={boxRef}>
                    <figure ref={logoRef}>
                        <img src={logo} alt="site logo" />
                    </figure>
                    <h1 ref={titleRef}>
                        <span>Reddit</span>Client
                    </h1>
                </section>
            </div>
        </header>
    )
}