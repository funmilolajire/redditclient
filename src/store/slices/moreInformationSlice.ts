import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { moreInformation } from '../api';

//set initial state
const initialState: MoreInformationState = {
    comments: [],
    postData: {},
    loading: true,
    permalink: '',
    showModal: false,
}

//get more information on post(i.e post information and comments) for modal
export const getMoreInformation = createAsyncThunk('moreInformation__modal', moreInformation)

//create more information slice
const moreInformationSlice = createSlice({
    name: "comments",
    initialState,
    reducers: {
        handleModalState: (state: MoreInformationState, { payload }: { payload: boolean }) => {
            state.loading = true
            state.showModal = payload;
            if (state.showModal === true) {
                document.getElementsByTagName('body')[0].classList.add('overflow-hidden');
            } else {
                document.getElementsByTagName('body')[0].classList.remove('overflow-hidden');
                document.getElementsByTagName('body')[0].classList.add('overflow-auto');
            }
        },
        handlePermalinkChange: (state: MoreInformationState, { payload }: { payload: string }) => {
            state.loading = true
            state.permalink = payload;
        },
        //for testing
        handleLoadingChange: (state: MoreInformationState, { payload }: { payload: boolean }) => {
            state.loading = payload;
        }
    },
    extraReducers: builder => {
        builder.addCase(getMoreInformation.fulfilled, (state: MoreInformationState, { payload }) => {
            if (payload) {
                payload.comments.pop()
                state.comments = payload.comments;
                state.postData = payload.postData;
                state.loading = false
            }
        }
        )
    }
})

//export more information reducer and actions
export default moreInformationSlice.reducer;
export const { handleModalState, handlePermalinkChange, handleLoadingChange } = moreInformationSlice.actions;