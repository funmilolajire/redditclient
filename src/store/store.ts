import { configureStore } from '@reduxjs/toolkit';
import termReducer from './slices/termSlice';

export default configureStore({
  reducer: {
    term: termReducer,
  },
});
