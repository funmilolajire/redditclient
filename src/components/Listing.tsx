import { useRef, useLayoutEffect, useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { handleListingChange, handleTermChange } from '../store/slices/postsSlice';

export const Listing = ({ listItem }: { listItem: string }) => {
    const listingRef = useRef<HTMLLIElement>(null);

    //handle listing change and active class
    const dispatch = useAppDispatch();
    const listing = useAppSelector(state => state.posts.listing);
    const handleListing = () => {
        if (null !== listingRef.current && listing !== listingRef.current.id) {
            let newListing = listingRef.current.id;
            dispatch(handleListingChange(newListing))
            dispatch(handleTermChange(''))
        }
    }
    const handleActiveClass = useCallback(() => {
        if (null !== listingRef.current) {
            const classList = listing === listingRef.current.id ? "Listing active" : "Listing";
            listingRef.current.className = classList
        }
    }, [listing])
    useLayoutEffect(() => {
        handleActiveClass();
    }, [handleActiveClass])


    //JSX
    return (
        <li ref={listingRef} data-testid={listItem} onClick={handleListing} id={listItem}>
            {listItem}
        </li>
    )
}