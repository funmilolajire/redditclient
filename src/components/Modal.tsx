import { useRef, useEffect, useLayoutEffect } from 'react';
import * as ReactDOM from 'react-dom';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { getMoreInformation, handleModalState } from '../store/slices/moreInformationSlice';
import { gsap } from 'gsap';
import { IoIosCloseCircleOutline } from 'react-icons/io';
import { ModalPost } from './Modal--Post';
import { ModalComments } from './Modal--Comments';

export const Modal = () => {
    //animations
    const modalRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([modalRef.current], {
            opacity: 0,
            x: '100vw',
            ease: 'circ.out',
            duration: 2
        })
    }, [])

    //handle modal
    const permalink = useAppSelector((state) => state.moreInformation.permalink)
    const closeModal = () => {
        dispatch(handleModalState(false))
    }

    //get post comments
    const dispatch = useAppDispatch()
    useEffect(() => {
        dispatch(getMoreInformation(permalink))
    }, [permalink, dispatch])

    //retrieve data from state for modal
    const postData = useAppSelector((state) => state.moreInformation.postData)
    const comments = useAppSelector((state) => state.moreInformation.comments)
    const loading = useAppSelector((state) => state.moreInformation.loading)

    //JSX Element
    return ReactDOM.createPortal(
        <div ref={modalRef} className="Modal">
            <div className="Modal--Inner">
                <header className="Modal--Header">
                    <span className="Modal--Close" onClick={closeModal}><IoIosCloseCircleOutline /></span>
                </header>
                {!loading ?
                    <div className="Modal--Content" data-testid="Modal--Content">
                        <>
                            <ModalPost post={postData} />
                            <ModalComments comments={comments} />
                        </>
                    </div> :
                    <div className="Modal--Loading">
                        <span></span>
                    </div>
                }
            </div>
        </div>
        , document.getElementById('portal') as HTMLElement)

}