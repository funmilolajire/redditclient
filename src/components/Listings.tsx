import { useRef, useLayoutEffect } from 'react';
import { Listing } from './Listing';
import { gsap } from 'gsap';

export const Listings = () => {
    const listings: string[] = ['hot', 'new', 'top', 'best', 'rising', 'controversial'];

    //animations
    const listingBoxRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([listingBoxRef.current], {
            x: '1200%',
            delay: 2,
            duration: 2,
            ease: 'power3.inOut'
        })
    }, [])

    return (
        <div ref={listingBoxRef} className="Listings">
            <ul className="Listings--List">
                {
                    listings && listings.map(listing => <Listing key={listing} listItem={listing} />)
                }
            </ul>
        </div>
    )
}