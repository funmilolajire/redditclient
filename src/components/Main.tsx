import { Listings } from './Listings';
import { Posts } from './Posts';
export const Main = () => (
    <div className="Main">
        <Listings />
        <Posts />
    </div>
)