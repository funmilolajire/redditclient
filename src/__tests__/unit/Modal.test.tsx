import { mount } from 'enzyme';
import { Modal } from '../../components/Modal';
import { Provider } from 'react-redux';
import { store } from '../../store/store';
import { handleLoadingChange } from '../../store/slices/moreInformationSlice';

describe('renders Modal component', () => {
    store.dispatch(handleLoadingChange(false))
    const portal = global.document.createElement('div')
    portal.setAttribute('id', 'portal')
    const body = global.document.querySelector('body')
    body?.appendChild(portal)
    let ModalComponent = mount(
        <Provider store={store}>
            <Modal />
        </Provider>
    )

    test('make a snapshot of Modal Component', () => {
        expect(ModalComponent).toMatchSnapshot()
    })

    test('Modal closes on click', () => {
        const closeModal = ModalComponent.find('.Modal--Close')
        closeModal.simulate('click');
        const modalState = store.getState().moreInformation.showModal;
        expect(modalState).toBeFalse();
    })
})