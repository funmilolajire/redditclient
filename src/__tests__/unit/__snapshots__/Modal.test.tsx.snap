// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Modal component make a snapshot of Modal Component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Modal>
    <Portal
      containerInfo={
        <div
          id="portal"
        >
          <div
            class="Modal"
            style=""
          >
            <div
              class="Modal--Inner"
            >
              <header
                class="Modal--Header"
              >
                <span
                  class="Modal--Close"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M331.3 308.7L278.6 256l52.7-52.7c6.2-6.2 6.2-16.4 0-22.6-6.2-6.2-16.4-6.2-22.6 0L256 233.4l-52.7-52.7c-6.2-6.2-15.6-7.1-22.6 0-7.1 7.1-6 16.6 0 22.6l52.7 52.7-52.7 52.7c-6.7 6.7-6.4 16.3 0 22.6 6.4 6.4 16.4 6.2 22.6 0l52.7-52.7 52.7 52.7c6.2 6.2 16.4 6.2 22.6 0 6.3-6.2 6.3-16.4 0-22.6z"
                    />
                    <path
                      d="M256 76c48.1 0 93.3 18.7 127.3 52.7S436 207.9 436 256s-18.7 93.3-52.7 127.3S304.1 436 256 436c-48.1 0-93.3-18.7-127.3-52.7S76 304.1 76 256s18.7-93.3 52.7-127.3S207.9 76 256 76m0-28C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                    />
                  </svg>
                </span>
              </header>
              <div
                class="Modal--Content"
                data-testid="Modal--Content"
              >
                <div
                  class="Modal--Post"
                >
                  <p
                    class="Post--Info"
                  >
                    posted 
                    <span
                      class="time"
                    />
                     by 
                    <span
                      class="user"
                    />
                     to 
                    <span
                      class="subreddit"
                    />
                  </p>
                  <div
                    class="Post--Content--Wrapper"
                  >
                    <div
                      class="Post--Content"
                    >
                      <h3
                        class="Post--Title"
                        data-read-more="Click Title to Open"
                      />
                      <section>
                        <div
                          class="Post--Bottom"
                        >
                          <div
                            class="Post--Comments"
                          >
                            <svg
                              class="comment-icon icon"
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              stroke-width="0"
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M20,2H4C2.897,2,2,2.897,2,4v18l5.333-4H20c1.103,0,2-0.897,2-2V4C22,2.897,21.103,2,20,2z M20,16H6.667L4,18V4h16V16z"
                              />
                            </svg>
                            <span
                              class="comments value"
                            />
                            <span
                              class="label"
                            >
                              comments
                            </span>
                          </div>
                          <svg
                            class="dot-icon"
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 16 16"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                              fill-rule="evenodd"
                            />
                          </svg>
                          <div
                            class="Votes Post--Upvotes"
                          >
                            <svg
                              class="arrow-icon icon"
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              stroke-width="0"
                              viewBox="0 0 10 16"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M5 3L0 9h3v4h4V9h3L5 3z"
                                fill-rule="evenodd"
                              />
                            </svg>
                            <span
                              class="upvotes value"
                            />
                            <span
                              class="label"
                            >
                              upvotes
                            </span>
                          </div>
                          <svg
                            class="dot-icon"
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 16 16"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                              fill-rule="evenodd"
                            />
                          </svg>
                          <div
                            class="Votes Post--Downvotes"
                          >
                            <svg
                              class="arrow-icon icon"
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              stroke-width="0"
                              viewBox="0 0 10 16"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M7 7V3H3v4H0l5 6 5-6H7z"
                                fill-rule="evenodd"
                              />
                            </svg>
                            <span
                              class="downvotes value"
                            />
                            <span
                              class="label"
                            >
                              downvotes
                            </span>
                          </div>
                        </div>
                      </section>
                    </div>
                  </div>
                </div>
                <div
                  class="Modal--Comments"
                >
                  <p
                    class="no--comments"
                  >
                    No comments yet
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <div
        className="Modal"
      >
        <div
          className="Modal--Inner"
        >
          <header
            className="Modal--Header"
          >
            <span
              className="Modal--Close"
              onClick={[Function]}
            >
              <IoIosCloseCircleOutline>
                <IconBase
                  attr={
                    Object {
                      "viewBox": "0 0 512 512",
                    }
                  }
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M331.3 308.7L278.6 256l52.7-52.7c6.2-6.2 6.2-16.4 0-22.6-6.2-6.2-16.4-6.2-22.6 0L256 233.4l-52.7-52.7c-6.2-6.2-15.6-7.1-22.6 0-7.1 7.1-6 16.6 0 22.6l52.7 52.7-52.7 52.7c-6.7 6.7-6.4 16.3 0 22.6 6.4 6.4 16.4 6.2 22.6 0l52.7-52.7 52.7 52.7c6.2 6.2 16.4 6.2 22.6 0 6.3-6.2 6.3-16.4 0-22.6z"
                      key="0"
                    />
                    <path
                      d="M256 76c48.1 0 93.3 18.7 127.3 52.7S436 207.9 436 256s-18.7 93.3-52.7 127.3S304.1 436 256 436c-48.1 0-93.3-18.7-127.3-52.7S76 304.1 76 256s18.7-93.3 52.7-127.3S207.9 76 256 76m0-28C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                      key="1"
                    />
                  </svg>
                </IconBase>
              </IoIosCloseCircleOutline>
            </span>
          </header>
          <div
            className="Modal--Content"
            data-testid="Modal--Content"
          >
            <ModalPost
              post={Object {}}
            >
              <div
                className="Modal--Post"
              >
                <p
                  className="Post--Info"
                >
                  posted 
                  <span
                    className="time"
                  />
                   by 
                  <span
                    className="user"
                  />
                   to 
                  <span
                    className="subreddit"
                  />
                </p>
                <div
                  className="Post--Content--Wrapper"
                >
                  <PostContent
                    post={Object {}}
                  >
                    <div
                      className="Post--Content"
                    >
                      <h3
                        className="Post--Title"
                        data-read-more="Click Title to Open"
                      />
                      <section>
                        <div
                          className="Post--Bottom"
                        >
                          <div
                            className="Post--Comments"
                          >
                            <BiComment
                              className="comment-icon icon"
                            >
                              <IconBase
                                attr={
                                  Object {
                                    "viewBox": "0 0 24 24",
                                  }
                                }
                                className="comment-icon icon"
                              >
                                <svg
                                  className="comment-icon icon"
                                  fill="currentColor"
                                  height="1em"
                                  stroke="currentColor"
                                  strokeWidth="0"
                                  style={
                                    Object {
                                      "color": undefined,
                                    }
                                  }
                                  viewBox="0 0 24 24"
                                  width="1em"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M20,2H4C2.897,2,2,2.897,2,4v18l5.333-4H20c1.103,0,2-0.897,2-2V4C22,2.897,21.103,2,20,2z M20,16H6.667L4,18V4h16V16z"
                                    key="0"
                                  />
                                </svg>
                              </IconBase>
                            </BiComment>
                            <span
                              className="comments value"
                            />
                            <span
                              className="label"
                            >
                              comments
                            </span>
                          </div>
                          <BsDot
                            className="dot-icon"
                          >
                            <IconBase
                              attr={
                                Object {
                                  "fill": "currentColor",
                                  "viewBox": "0 0 16 16",
                                }
                              }
                              className="dot-icon"
                            >
                              <svg
                                className="dot-icon"
                                fill="currentColor"
                                height="1em"
                                stroke="currentColor"
                                strokeWidth="0"
                                style={
                                  Object {
                                    "color": undefined,
                                  }
                                }
                                viewBox="0 0 16 16"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clipRule="evenodd"
                                  d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                                  fillRule="evenodd"
                                  key="0"
                                />
                              </svg>
                            </IconBase>
                          </BsDot>
                          <div
                            className="Votes Post--Upvotes"
                          >
                            <GoArrowUp
                              className="arrow-icon icon"
                            >
                              <IconBase
                                attr={
                                  Object {
                                    "viewBox": "0 0 10 16",
                                  }
                                }
                                className="arrow-icon icon"
                              >
                                <svg
                                  className="arrow-icon icon"
                                  fill="currentColor"
                                  height="1em"
                                  stroke="currentColor"
                                  strokeWidth="0"
                                  style={
                                    Object {
                                      "color": undefined,
                                    }
                                  }
                                  viewBox="0 0 10 16"
                                  width="1em"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M5 3L0 9h3v4h4V9h3L5 3z"
                                    fillRule="evenodd"
                                    key="0"
                                  />
                                </svg>
                              </IconBase>
                            </GoArrowUp>
                            <span
                              className="upvotes value"
                            />
                            <span
                              className="label"
                            >
                              upvotes
                            </span>
                          </div>
                          <BsDot
                            className="dot-icon"
                          >
                            <IconBase
                              attr={
                                Object {
                                  "fill": "currentColor",
                                  "viewBox": "0 0 16 16",
                                }
                              }
                              className="dot-icon"
                            >
                              <svg
                                className="dot-icon"
                                fill="currentColor"
                                height="1em"
                                stroke="currentColor"
                                strokeWidth="0"
                                style={
                                  Object {
                                    "color": undefined,
                                  }
                                }
                                viewBox="0 0 16 16"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clipRule="evenodd"
                                  d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
                                  fillRule="evenodd"
                                  key="0"
                                />
                              </svg>
                            </IconBase>
                          </BsDot>
                          <div
                            className="Votes Post--Downvotes"
                          >
                            <GoArrowDown
                              className="arrow-icon icon"
                            >
                              <IconBase
                                attr={
                                  Object {
                                    "viewBox": "0 0 10 16",
                                  }
                                }
                                className="arrow-icon icon"
                              >
                                <svg
                                  className="arrow-icon icon"
                                  fill="currentColor"
                                  height="1em"
                                  stroke="currentColor"
                                  strokeWidth="0"
                                  style={
                                    Object {
                                      "color": undefined,
                                    }
                                  }
                                  viewBox="0 0 10 16"
                                  width="1em"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M7 7V3H3v4H0l5 6 5-6H7z"
                                    fillRule="evenodd"
                                    key="0"
                                  />
                                </svg>
                              </IconBase>
                            </GoArrowDown>
                            <span
                              className="downvotes value"
                            />
                            <span
                              className="label"
                            >
                              downvotes
                            </span>
                          </div>
                        </div>
                      </section>
                    </div>
                  </PostContent>
                </div>
              </div>
            </ModalPost>
            <ModalComments
              comments={Array []}
            >
              <div
                className="Modal--Comments"
              >
                <p
                  className="no--comments"
                >
                  No comments yet
                </p>
              </div>
            </ModalComments>
          </div>
        </div>
      </div>
    </Portal>
  </Modal>
</Provider>
`;
